datarobot_batch_scoring/detect.py

```diff
@@ -0,0 +1,91 @@
+import math
+import re
+
+
+class Detector(object):
+
+    frequency_table = None
+    non_delimiter_re = re.compile('[\w]', re.UNICODE)
+    buff_size = 1024 * 2048
+
+    def __init__(self):
+        self.frequency_table = {}
+
+    def increment(self, char, lines):
+        if char not in self.frequency_table:
+            self.frequency_table[char] = {}
+        if lines not in self.frequency_table[char]:
+            self.frequency_table[char][lines] = 0
+
+        self.frequency_table[char][lines] += 1
+
+    def detect(self, sample):
+        lines_analyzed, lines_content = self.get_sample(sample)
+        resampled = "".join(lines_content[:-1])
+        return self.analyze(lines_analyzed), resampled
+
+    def get_sample(self, sample, sample_lines=20, quotechar='"'):
+        enclosed = False
+        actual_lines = 1
+        lines = []
+        single_line = ""
+
+        # raise Exception("Sample size %d" % len(sample))
+
+        for idx, ch in enumerate(sample):
+            if actual_lines >= sample_lines:
+                break
+            prev = sample[idx-1] if idx else None
+            next = sample[idx+1] if (idx + 1) < len(sample) else None
+            single_line += ch
+
+            if ch == quotechar:
+                if enclosed and next != quotechar:
+                    enclosed = False
+                elif not enclosed:
+                    enclosed = True
+            elif not enclosed and (ch == '\n' and prev != '\r' or ch == '\r'):
+                actual_lines += 1
+                lines.append(single_line)
+                single_line = ""
+
+            elif not enclosed and not self.non_delimiter_re.match(ch):
+                self.increment(ch, actual_lines)
+
+        return actual_lines, lines
+
+    def analyze(self, lines_analyzed):
+        candidates = []
+        for delim, freq in list(self.frequency_table.items()):
```

Comment:
I thought items() returning iterable in both cases, and we don't need to list evaluate it.